exp/ipam/internal/webhooks/ipaddress.go

```diff
@@ -0,0 +1,162 @@
+
+// IPAddress implements a validating webhook for IPAddress.
+type IPAddress struct {
+	Client client.Reader
+}
+
+var _ webhook.CustomValidator = &IPAddress{}
+
+// ValidateCreate implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	ip, ok := obj.(*ipamv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", obj))
+	}
+	return webhook.validate(ctx, ip)
+}
+
+// ValidateUpdate implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	oldIP, ok := oldObj.(*ipamv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", oldObj))
+	}
+	newIP, ok := newObj.(*ipamv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", newObj))
+	}
+
+	if !reflect.DeepEqual(oldIP.Spec, newIP.Spec) {
+		return field.Forbidden(field.NewPath("spec"), "the spec of IPAddress is immutable")
+	}
+	return nil
+}
+
+// ValidateDelete implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	return nil
+}
+
+func (webhook *IPAddress) validate(ctx context.Context, ip *ipamv1.IPAddress) error {
+	log := ctrl.LoggerFrom(ctx)
+	allErrs := field.ErrorList{}
+
+	addr, err := netaddr.ParseIP(ip.Spec.Address)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.address"),
+				ip.Spec.Address,
+				"not a valid IP address",
+			))
+	}
+
+	if ip.Spec.Prefix < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix cannot be negative",
+			))
+	}
+	if addr.Is4() && ip.Spec.Prefix > 32 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv4 address",
+			))
+	}
+	if addr.Is6() && ip.Spec.Prefix > 128 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv6 address",
+			))
+	}
+
+	_, err = netaddr.ParseIP(ip.Spec.Gateway)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.gateway"),
+				ip.Spec.Gateway,
+				"not a valid IP address",
+			))
+	}
+
+	if ip.Spec.PoolRef.APIGroup == nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.poolRef.apiGroup"),
+				ip.Spec.PoolRef.APIGroup,
+				"the pool reference needs to contain a group"))
+	}
+
+	claim := &ipamv1.IPAddressClaim{}
+	err = webhook.Client.Get(ctx, types.NamespacedName{Name: ip.Spec.ClaimRef.Name, Namespace: ip.ObjectMeta.Namespace}, claim)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to fetch claim", "name", ip.Spec.ClaimRef.Name)
```

Comment:
I think it's okay to append this error to allErrs as well. We don't hit this case in race conditions / ordering issues, just when we have some other problem with the APIserver (If I understood it correctly)